plugins/zenoh-plugin-storage-manager/src/storages_mgt/service.rs

```diff
@@ -470,25 +491,32 @@ impl StorageService {
     async fn guard_cache_if_latest(
         &self,
         stripped_key: &Option<OwnedKeyExpr>,
-        timestamp: &Timestamp,
-    ) -> Option<MutexGuard<'_, HashMap<Option<OwnedKeyExpr>, Timestamp>>> {
-        let cache_guard = self.latest_updates.lock().await;
-        if let Some(latest_timestamp) = cache_guard.get(stripped_key) {
-            if timestamp > latest_timestamp {
+        received_ts: &Timestamp,
+    ) -> Option<RwLockWriteGuard<'_, LatestUpdates>> {
+        let cache_guard = self.cache_latest.latest_updates.write().await;
+        if let Some(event) = cache_guard.get(stripped_key) {
+            if received_ts > event.timestamp() {
                 return Some(cache_guard);
-            } else {
-                return None;
             }
         }
 
-        let mut storage = self.storage.lock().await;
-
-        if let Ok(stored_data) = storage.get(stripped_key.clone(), "").await {
-            for data in stored_data {
-                if data.timestamp > *timestamp {
+        if let Some(replication_log) = &self.cache_latest.replication_log {
+            if let Some(event) = replication_log.read().await.lookup(stripped_key) {
+                if received_ts <= event.timestamp() {
                     return None;
                 }
             }
+        } else {
+            let mut storage = self.storage.lock().await;
+            // FIXME: An actual error from the underlying Storage cannot be distinguished from a
+            //        missing entry.
```

Comment:
Which storage produce these errors ? 
I can open issues in the storages to fix.

Reply:
In my opinion the problem lies in this method: https://github.com/eclipse-zenoh/zenoh/blob/ddcc8f1d8c114ca256df3939c303bcbdc1f80197/plugins/zenoh-backend-traits/src/lib.rs#L247-L251

It should return an `ZResult<Option<Vec<StoredData>>>` where the `None` indicates that there is no entry for that key expression.

Right now, for instance, the Memory backend returns an error: https://github.com/eclipse-zenoh/zenoh/blob/main/plugins/zenoh-plugin-storage-manager/src/memory_backend/mod.rs#L131

I think this deserve a dedicated issue and discussion.